_posts/2019-07-30-indexing-searching-nuget-with-azure-functions-and-search.markdown

```diff
@@ -0,0 +1,891 @@
+
+Now, between 2015 and now a number of things happened... First, increased NuGet usage and an increasing number of NuGet packages! The daily upload counts were increasing, and our indexer kept receiving more and more work.
+
+<blockquote class="twitter-tweet"><p lang="en" dir="ltr">huh, <a href="https://t.co/tn7WCIynaN">nuget.org</a> repo is 1.9Tb now... was like 250Gb in a year 2015 <a href="https://t.co/aP73rbgq9J">pic.twitter.com/aP73rbgq9J</a></p>&mdash; Alexander Shvedov (@controlflow) <a href="https://twitter.com/controlflow/status/1067724815958777856?ref_src=twsrc%5Etfw">November 28, 2018</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>
+
+Second, over the 2018/2019 holiday season, [NuGet started repository-signing packages](https://blog.nuget.org/20180810/Introducing-Repository-Signatures.html). A good thing, with one caveat: during that period every single package that had been on NuGet.org was updated (a new editing timestamp, plus a fresh binary with that repository signature attached). In other words, sort of a complete re-index was happening on our end.
+
+Not a big issue, except that re-index was becoming very slow. The OData feed exposed by NuGet is essentially "SQL over HTTP", where we were receiving 100 packages per page of results, thus running `# packages / 100` order by queries against their database.
+
+Could there be a better way to fetch package metadata?
+
+### NuGet API v3
+
+Turns out there *is* a better way to fetch package metadata. As we all know, NuGet talks to a repository - either a folder or network share, or a remote one over HTTP.
+
+Those HTTP(S) based ones have been evolving over time. NuGet.org, as well as other NuGet server implementations out there support either one or both of the supported HTTP API's:
+
+* V2 API – OData based (used by pretty much all NuGet servers out there)
+* V3 API – JSON based (NuGet.org, TeamCity, MyGet, Azure DevOps, GitHub repos, Sleet, ...)
+
+That V3 API is not really "one" API. It's a set of API's that are made available, for different use cases. If you look at [NuGet.org's V3 API URL](https://api.nuget.org/v3/index.json), we will see a number of "resources" listed:
+
+* V2 API URL
+* Package upload URL
+* V3 search URL
+* V3 autocompletion URL
+* Registrations (materialization of latest metadata state of a package)
+* Flat container (.NET Core package restore and basic version metadata)
+* Catalog (append-only event log, NuGet.org only)
+* Report abuse URL template
+* ...
+
+The NuGet client uses these depending on the scenario. Obviously, it uses search when searching packages, but will use the flat container for package restore as that API is much better suited for quickly resolving package binaries.
+
+### NuGet catalog
```

Comment:
I'm a fan of capitalized titles, but up to you
```suggestion
### NuGet Catalog
```

Reply:
I'm not :-) (even though there are some conventions out there :-))

Reply:
That's fine!